City-of-Helsinki/drupal-helfi-etusivu
# Uhf 6036 json to drupal menu block

# [Base block for JSON -> Menu conversion](https://helsinkisolutionoffice.atlassian.net/browse/UHF-6036)
This branch adds base block for creating menus from JSON input. This is to be used in the upcoming global navigation feature.

## What was done
Added: 
- JSON schema and validation for global menu JSON
- Added new supporint classes to create 'external menu trees'
- Added factory class for creating external menu trees
- Added base block to be extended into complete menu block
- Added interface for exernal menus that should be implemented by external menu blocks

## How to install

- Run `make fresh` or `make new`
- Run `drush en helfi_global_navigation -y` (note: module is not enabled on purpose, we don't want to enable it in production yet)

## How to test
Since what this feature creates is a base functionality, we need to add some files to actually test it.
Add this file: `public/modules/custom/helfi_global_navigation/assets/test.json`
<details>
<summary>..and add this 👇 inside it</summary>

```
[
  {
    "id": "123123",
    "name": "Etusivu",
    "url": "https://etusivu.docker.so",
    "weight": 0,
    "menu_tree": [
      {
        "id": "1234",
        "name": "City of Helsinki - Etusivu",
        "url": "https://etusivu.docker.so",
        "weight":0,
        "menu_tree": [
          {
            "id": "12345",
            "description": "Ei oikea sivu",
            "name":"City of Helsinki - Takasivu",
            "url": "https://etusivu.docker.so",
            "weight":0,
            "menu_tree": [
              {
                "id": "12323423445",
                "description": "NELJÄS TASO",
                "name":"City of Helsinki - NELJÄS TASO",
                "url": "https://etusivu.docker.so",
                "weight":0
              }
            ]
          }
        ]
      },
      {
        "id": "123456",
        "name": "City of Helsinki - Etusivu",
        "url": "https://etusivu.docker.so",
        "weight":0
      }
    ]
  },
  {
    "id": "213",
    "name": "Sote",
    "url": "https://helfi-sote.docker.so",
    "menu_tree": [
      {
        "id": "214",
        "name": "City of Helsinki - Sote",
        "url":"https://helfi-sote.docker.so",
        "weight":0
      }
    ]
  }
]

```
</details>

Then, add `public/modules/helfi_global_navigation/Plugin/Block/TestBlock.php`
<details>
<summary>..and add this 👇 inside it</summary>

```
<?php

namespace Drupal\helfi_global_navigation\Plugin\Block;

use Drupal\helfi_global_navigation\ExternalMenuBlockInterface;

/**
 * For testing purposes.
 *
 * @Block(
 *   id = "test_block",
 *   admin_label = @Translation("Test"),
 *   category = @Translation("Testing")
 * )
 */
class TestBlock extends ExternalMenuBlockBase implements ExternalMenuBlockInterface {

  /**
   * {@inheritdoc}
   */
  public function getData(): string {
    $path = __DIR__ . '/../../../assets/test.json';

    return file_get_contents(__DIR__ . '/../../../assets/test.json');
  }

  /**
   * {@inheritdoc}
   */
  public function maxDepth(): int {
    return 2;
  }

}
```
</details>

... then go to [block layout administration](https://helfi-etusivu.docker.so/fi/admin/structure/block) and add the "test" block to a region (I do recommend content-region, to make viewing the data easy). 

Now that that is done, testing can be done with:

- Run `drush cr`
- View a page. You should see the data that was inside test.json printed out throught the menu template.
- Try 'breaking' the json by deleting id from one of the menu items. The menu should not load after clearing cache.
- Menu items should get ordered by weight value (higher weight = lower in the menu). You can test this by editing the test.json file. 


## Designers review
